### PR TITLE
Custom functions and filters

### LINKED ISSUE #8: Allow to register custom filters & functions

### PR DESCRIPTION
**Pull request by zrzka**
Adds ability to register custom functions and filters.

Fixes #8

### CODE REVIEW

File: tests/engine/eval/filter.rs [hunk 1]
```diff
@@ -43,3 +43,24 @@ fn test_filter_chain() {
...
+        if value.is_string() {
+            Ok(Value::String(value.as_str().unwrap().replace("a", "b")))
+        } else {
+            Err("no string, no fun".into())
+        }
+    };
+
+    let mut engine = Engine::default();
+    engine.register_filter("atob", cf);
+    let ctx = Context::default();
```
Comment (cyplo) [inline tests/engine/eval/filter.rs]: from the point of view of the user - in this case - do they need to know about `context` ? Would it be possible/make sense to have a interface where they just do `eval` ?

File: src/engine/mod.rs [hunk 2]
```diff
@@ -22,14 +22,58 @@ pub struct Engine {
 
 impl Default for Engine {
     fn default() -> Engine {
-        let mut filters = HashMap::new();
-        filters.insert("upper".into(), Box::new(filter::upper) as FilterFn);
-        filters.insert("lower".into(), Box::new(filter::lower) as FilterFn);
+        let mut engine = Engine::new();
+        engine.register_builtin_filters();
```
Comment (cyplo) [inline src/engine/mod.rs]: Nit - would it make sense to move the `register_builtin...` calls to a separate constructor, like `with_default_builtins` and then call it from here ?

Comment (cyplo) [reply src/engine/mod.rs]: More like a thought here - would it be more work if we made the engine immutable and the `register...` methods return a new engine ? I found that it typically leads to clearer user code if things are externally immutable.

File: tests/engine/eval/filter.rs (hunk 1 above)
Comment (zrzka) [reply tests/engine/eval/filter.rs]: Yes, they must know about it. `context` contains DRY json (data) and will contain other things in the future.

File: src/engine/mod.rs (hunk 2 above)
Comment (zrzka) [reply src/engine/mod.rs]: Nope, there's `Default` and it contains all built ins. `new` is private, user shouldn't create empty engine (without builtins).

Comment (zrzka) [reply src/engine/mod.rs]: Re immutability - then I'd introduce something like `EngineBuilder`. I'll do it to show to you what I mean. Give me few minutes.

**Pull request merged**